src/scheduler/MpiWorld.cpp

```diff
@@ -912,53 +918,138 @@ void MpiWorld::gather(int sendRank,
-                // position
-                continue;
-            } else if (r == recvRank) {
-                // Copy data locally on root
-                std::copy(sendBuffer, sendBuffer + sendOffset, recvChunk);
+        for (auto it : ranksForHost) {
+            if (it.first == thisHost) {
+                // Each local rank will send a separate message that we receive
+                // directly in the receive buffer.
+                for (const int r : it.second) {
+                    if (r == recvRank) {
+                        // If not receiving in place, copy our data to the
+                        // right offset
+                        if (!isInPlace) {
+                            memcpy(recvBuffer + (recvRank * recvOffset),
+                                   sendBuffer,
+                                   sendOffset);
+                        }
+
+                        continue;
+                    }
+
+                    recv(r,
+                         recvRank,
+                         recvBuffer + (r * recvOffset),
+                         recvType,
+                         recvCount,
+                         nullptr,
+                         faabric::MPIMessage::GATHER);
+                }
             } else {
-                // Receive data from rank if it's not the root
-                recv(r,
+                // Each remote leader will send a signle message with the data
+                // for all its local ranks
+                auto rankData =
+                  std::make_unique<uint8_t[]>(it.second.size() * recvOffset);
+
+                recv(it.second.front(),
                      recvRank,
-                     recvChunk,
+                     rankData.get(),
                      recvType,
-                     recvCount,
+                     recvCount * it.second.size(),
                      nullptr,
                      faabric::MPIMessage::GATHER);
+
+                // Copy each received chunk to its offset
+                for (int r = 0; r < it.second.size(); r++) {
+                    memcpy(recvBuffer + (it.second.at(r) * recvOffset),
+                           rankData.get() + (r * recvOffset),
+                           recvOffset);
+                }
+            }
+        }
+    } else if (sendRank == localLeader) {
+        if (getHostForRank(recvRank) != thisHost) {
+            // If we are the local leader (but not the receiver of the gather)
+            // and the receiver is not co-located with us, we gather all our
+            // local rank's data and send it in a single message
+            auto rankData = std::make_unique<uint8_t[]>(
+              ranksForHost[thisHost].size() * sendOffset);
+
+            for (int r = 0; r < ranksForHost[thisHost].size(); r++) {
+                if (ranksForHost[thisHost].at(r) == sendRank) {
+                    // If gathering in-place from a rank that is not-receiving
```

Comment:
Gather has an added factor of grimness because the offset were the data to be sent lives in the send buffer depends:
* If doing an `MPI_Allgather` with `MPI_IN_PLACE`, the send buffer already contains space for the gather result, and the send data is at the offset corresponding to the sending rank.
* In any other situation it is directly at offset zero.